fix(ivy): don't load factories when using Ivy

filipesilva opened the pull request:
Related to https://github.com/angular/angular/pull/26947/files#r251221208
Related to https://github.com/angular/angular-cli/pull/13524

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mary-poppins commented:
You can preview 2eba24e at https://pr28685-2eba24e.ngbuilds.io/.

filipesilva commented:
Related and probably blocked on https://github.com/angular/angular-cli/pull/13667

mary-poppins commented:
You can preview 8168a38 at https://pr28685-8168a38.ngbuilds.io/.

brandonroberts commented:
Also needs to be removed from here https://github.com/angular/angular/blob/master/aio/tools/examples/shared/boilerplate/ivy/cli/src/tsconfig.app.json

filipesilva committed:
fix(ivy): don't load factories when using Ivy

Related to https://github.com/angular/angular/pull/26947/files#r251221208
Related to angular/angular-cli#13524

filipesilva commented on aio/package.json:
```diff
@@ -97,7 +97,7 @@
   },
   "devDependencies": {
-    "@angular-devkit/build-angular": "^0.12.1",
+    "@angular-devkit/build-angular": "github:angular/angular-devkit-build-angular-builds#1bba17f",
```
Needs to be updated to `0.14.0-beta.1` when it is released.

filipesilva commented:
@brandonroberts thanks for the heads up, that's done now.

mary-poppins commented:
You can preview 0ff2b72 at https://pr28685-0ff2b72.ngbuilds.io/.

mary-poppins commented:
You can preview 17ca113 at https://pr28685-17ca113.ngbuilds.io/.

mary-poppins commented:
You can preview 45462bd at https://pr28685-45462bd.ngbuilds.io/.

filipesilva commented on packages/core/src/linker/system_js_ng_module_factory_loader.ts:
```diff
@@ -56,8 +57,8 @@ export class SystemJsNgModuleLoader implements NgModuleFactoryLoader {
   load(path: string): Promise<NgModuleFactory<any>> {
-    const offlineMode = this._compiler instanceof Compiler;
-    return offlineMode ? this.loadFactory(path) : this.loadAndCompile(path);
+    const legacyOfflineMode = !ivyEnabled && this._compiler instanceof Compiler;
```
I'm not sure this is the best way to address lazy loaded Ivy modules. 

This way makes it use `this.loadAndCompile` when in Ivy mode, which is just calling `NgModuleFactoryR3` in Ivy.

Other approaches include making a separate Ivy only SystemJsNgModuleLoader , or sending in an option in `SystemJsNgModuleLoaderConfig`.

@kara @alxhub can you advise?

IgorMinar replied:
I think this change is not right.

You should either enable factory shim generation OR not use NgModuleFactoryLoader.

But nobody should disable (or not enable) shimming and still try to use NgModuleFactoryLoader.

The reasoning being that if you don't want shims, then you should just use `import()` to lazy load code and not use our custom loaders that increase the complexity of the code.

Projects that want to be able to switch back and forth between the legacy and ivy pipelines should enable shims and use NgModuleFactoryLoader until they permanently  switch over to Ivy.

IgorMinar replied:
I think that the only other option we should entertain is that in ivy we would load the path and extract the factory from the NgModule loaded via `this.loadFactory(path)` - but I'm not sure if we can do that reliably.. maybe we can??? in that case one could switch back and forth between ivy and legacy without having to enable factory shimming....

filipesilva replied:
> I think that the only other option we should entertain is that in ivy we would load the path and extract the factory from the NgModule loaded via `this.loadFactory(path)`

Maybe that could be done. But it would also require tooling (like the CLI, or others) to understand that there is another difference in behaviour for the `loadChildren/NgModuleFactoryLoader` interaction.

Let's list the existing interactions for clarity though. I also tested all of these with the latest snapshots and without any of the changes in this PR.

Assuming the following `loadChildren`:
```
    RouterModule.forRoot([{
      path: 'lazy',
      loadChildren: './lazy.module#LazyModule',
    }]),
```

The VE ones are:
- without AOT, NgModuleFactoryLoader tries to load `lazy.module.js` (works now)
- with AOT, NgModuleFactoryLoader tries to load `lazy.module.ngfactory.js` (works now)

With Ivy and `allowEmptyCodegenFiles: true` it should be the same as above:
- without AOT, NgModuleFactoryLoader tries to load `lazy.module.js` 
  - **does not work**, `loadFactory` is used but there is no `.ngfactory.js`
- with AOT, NgModuleFactoryLoader tries to load `lazy.module.ngfactory.js` (works now)

With Ivy and `allowEmptyCodegenFiles: false`:
- if `loadChildren: './lazy.module#LazyModule',` is replaced with `loadChildren: () => import('./lazy/lazy.module').then(m => m.LazyModule),`, then `NgModuleFactoryLoader` is not used at all (works now).

All of these interactions work with no changes today, with the exception of jit+ivy+`allowEmptyCodegenFiles: true` (which should be fixed).

The novel one you are proposing is Ivy and `allowEmptyCodegenFiles: false`, with `loadChildren: './lazy.module#LazyModule',`:
- without AOT, NgModuleFactoryLoader tries to load `lazy.module.js`
  - **does not work**, `loadFactory` is used but there is no `.ngfactory.js`
- with AOT, NgModuleFactoryLoader tries to load `lazy.module.ngfactory.js`
  - **does not work**, `loadFactory` is used but there is no `.ngfactory.js`

I think the new case is possible to implement. 

But supporting it implies there is still a magic `import` call in the angular/core codebase that must be special cased by tooling, which is used whenever `loadChildren` has a string argument. 

Angular CLI currently special cases the `System.import()` call within `NgModuleFactoryLoader` by:
- calling a private API in `compiler-cli` that lists lazy routes
- assembling a map of lazy routes to file names
- informing the bundler that those file names can be used by the magic `import`

All of this adds to the rebuild times and requires maintenance. Not supporting the `loadChildren` string format would thus make builds faster, require less maintenance on our code, and require no special casing on third party tooling. So it sounds like a better long-term strategy to not have it.

The spanner in this approach is that `allowEmptyCodegenFiles: true` also provides support the `loadChildren` string format in Ivy. So while that mode exists it must be supported.

filipesilva replied:
Added tests for all these Ivy lazy load variations in https://github.com/angular/angular-cli/pull/13774.

filipesilva replied:
I am looking at ways to support the `allowEmptyCodegenFiles` variations in Ivy but think there's a fundamental problem with that approach: `allowEmptyCodegenFiles` is a compile-time option but using `SystemJsNgModuleLoader` is runtime behaviour, and I'm not sure there is a way to transfer that information over.

As far as I can tell `SystemJsNgModuleLoader` is always provided by `@angular/router`. 

The injected `Compiler` instance that `SystemJsNgModuleLoader` uses is provided by `@angular/core`. In VE, it is overriden in `@angular/platform-browser-dynamic`, but not in Ivy. So Ivy always uses the `@angular/core` `Compiler`.

`SystemJsNgModuleLoader` checks what kind `Compiler` to determine whether it should try to load `.ngfactory.js` files. If it has the `@angular/core` one, it loads them, otherwise it doesn't.

In Ivy it has the same `Compiler` in JIT and AOT, so always tries to load `.ngfactory.js`.

In VE the what determined the existence of `.ngfactory.js` files was JIT vs AOT. The way that was communicated to the runtime Angular execution was through the used platform: `platform-browser-dynamic` meant JIT, and `platform-browser` meant AOT.

But in Ivy what determines the existence of factories is not just JIT vs AOT, but rather an additional AOT-only compile-time option: `allowEmptyCodegenFiles: true`. There is currently no way to communicate this option to the runtime, thus it cannot determine if `.ngfactory.js` files should be loaded.

This leads to a couple of choices:

If we want `allowEmptyCodegenFiles: true` to serve as a switch within Ivy to support loading `.ngfactory.js` files, then that information must be provided to the runtime.

If we want `allowEmptyCodegenFiles: true` to generate `.ngfactory.js` files but not load them, then in Ivy `SystemJsNgModuleLoader` should not be provided.

If we want to support the `loadChildren` string format in Ivy, a `NgModuleFactoryLoader` must be provided that contains an `import` statement. This `import` statement must be special-cased by tooling.

filipesilva replied:
Another advantage of only supporting the `loadChildren` `import()` format in Ivy is that tooling TS configuration becomes simplified. 

We can point the `tsconfig` `files` property at `main.ts` and TS will figure out that `import()` calls include those files as well. This removes the complicated `include`/`exclude` setups we have.

sod replied:
So much this. Looking at the flamechart on recompile in our project with ivy i'm so confused:

![image](https://user-images.githubusercontent.com/905328/53827478-78b27780-3f7b-11e9-9472-97cbf241fcab.png)

I thought the whole premise of ivy was that it compiles file2file and doesn't need the knowledge of it's surroundings, that currently makes the build so slow and complicated. And lazy loading is a solved problem in webpack and very fast.

filipesilva replied:
@sod I'm not really sure what you're trying to highlight in that flamechart. I don't see anything about lazy routes there.

IgorMinar commented on packages/core/test/linker/system_ng_module_factory_loader_spec.ts:
```diff
@@ -54,4 +55,26 @@ function mockSystem(modules: {[module: string]: any}) {
+    it('loads a named module', async(() => {
+         const loader = new SystemJsNgModuleLoader(new Compiler());
+         loader.load('test#NamedModule').then(contents => {
+           expect(contents.moduleType).toBe('test NamedModule' as any);
```
this is not right.. the loaded type should have been of type NgModuleFactory extracted  out of the ES module and class loaded via `test#NamedModule`. This test proves that we are not extracting the factory but simply returning the loaded class instead and that's not right.

alxhub replied:
I'm not sure if the code here has changed since this comment, but the test looks correct to me. We're testing that `load` has created the `NgModuleFactory` wrapper, and that the `moduleType` (the original "class" that was dynamically loaded) is the one we expected it to wrap.

IgorMinar reviewed (dismissed):
this is almost good.. :-) 

please see my comments about the issues I think I found. Once those are resolved, the rest looks great.

IgorMinar commented:
Discussed this in person with @filipesilva, he's going to do some more exploration of supporting the `import()` in the legacy (ngc+viewengine) pipeline.

If that's feasible then we'll implement this in a way where in Ivy we'll *always* require the `import()` syntax to be used. `NgModuleFactoryLoader` api would be not work in Ivy and we'd deprecate it and removed in the future in favor of the standard `import()` api.

filipesilva commented:
@IgorMinar as far as I can tell there are two things stopping users from using `import()` in `loadChildren` in View Engine:
- `LoadChildrenCallback` works with `Promise<NgModuleFactory<any>>` but that's not listed in the typings  (fix in https://github.com/angular/angular/pull/29392)
- Angular CLI does not transform imports from the class to the ngfactory (fix in https://github.com/angular/angular-cli/pull/13948)

With both of those PRs in, the following syntax will work in ViewEngine and Ivy, both for JIT and AOT:
```js
const routes = [{
  path: 'lazy',
  loadChildren: () => import('./lazy/lazy.module').then(m => m.LazyModule)
}];    
```

Please see https://github.com/angular/angular-cli/pull/13948 for limitations of this transform.

alexeagle commented:
According to @alxhub we ought to move this ahead and land it. This would give us the option of supporting loadChildren:string in Ivy, which would be good if we find that the migration to dynamic import breaks users. @filipesilva could you rebase and clean up the unneeded version bumps?

mary-poppins commented:
You can preview be21c3f at https://pr28685-be21c3f.ngbuilds.io/.

mary-poppins commented:
You can preview f66bc93 at https://pr28685-f66bc93.ngbuilds.io/.

mary-poppins commented:
You can preview 1a31e55 at https://pr28685-1a31e55.ngbuilds.io/.

filipesilva committed:
test(docs-infra): ignore lazy loading examples in Ivy

mary-poppins commented:
You can preview 4d37a06 at https://pr28685-4d37a06.ngbuilds.io/.

kara commented:
[new presubmit](http://test/OCL:243299116:BASE:243299172:1555093267246:db600b95)
